### ISSUE TITLE
netopeer2-server Segmentation fault

### ISSUE THREAD

**Issue by jiantuo** [CESNET/netopeer2#711]
Hi Guys,

Just met a netopeer2-server Segmentation fault issue, could you please spare some time to help.  

We do use an ancient version of libs for some reason which beyond our control.
It maybe a issue not been found yet for the legacy branch.

Versions: 
       libyang:       version 0.15.120
       libnetconf2: version 0.11.37
       sysrepo:       version 0.7.4
       Netopeer2:  version 0.5.22

```VERBOSE: Session 1 received a realtime notification "/ietf-netconf-notifications:netconf-session-start" (1599036391).                                   
VERBOSE: Generated new event (netconf-session-start).                                                                                                   
VERBOSE: Session 1 received a realtime notification "/pnf-mplane-ext:pnf-restart-notification" (1599036403).                                            
[2020/09/02 08:46:43.548352, 1] ssh_socket_exception_callback:  Socket exception callback: 1 (0)                                                        
[2020/09/02 08:46:43.548398, 1] ssh_socket_exception_callback:  Socket error: disconnected                                                              
ERROR: Session 2: SSH channel poll error (Socket error: disconnected).                                                                                  
VERBOSE: Session 2: thread 3 event session terminated.                                                                                                  
ERROR: Session 1: SSH channel poll error (Socket error: disconnected).                                                                                  
VERBOSE: Session 1: thread 1 event session terminated.                                                                                                  
VERBOSE: Session 1 received a realtime notification "/ietf-netconf-notifications:netconf-session-end" (1599036403).                                     
VERBOSE: Generated new event (netconf-session-end).                                                                                                     
                                                                                                                                                        
Thread 5 "netopeer2-serve" received signal SIGSEGV, Segmentation fault.                                                                                 
[Switching to Thread 0x7f7ee97fa700 (LWP 464)]                                                                                                          
__GI___pthread_mutex_lock (mutex=0x0) at ../nptl/pthread_mutex_lock.c:65                                                                                
65      ../nptl/pthread_mutex_lock.c: No such file or directory.                                                                                        
(gdb) bt                                                                                                                                                
#0  __GI___pthread_mutex_lock (mutex=0x0) at ../nptl/pthread_mutex_lock.c:65                                                                            
#1  0x00007f7eeefe114b in nc_session_free (session=0x7f7ed8004470, data_free=0x564093c6f9d3 <free_ds>) at /opt/dev/libnetconf2.11/src/session.c:622     
#2  0x0000564093c70e88 in np2srv_del_session_clb (session=0x7f7ed8004470) at /opt/dev/Netopeer2/server/main.c:742                                       
#3  0x0000564093c72254 in worker_thread (arg=0x5640942fe4f0) at /opt/dev/Netopeer2/server/main.c:1180                                                   
#4  0x00007f7eef5026db in start_thread (arg=0x7f7ee97fa700) at pthread_create.c:463                                                                     
#5  0x00007f7eee90388f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95                                                                      
(gdb) info locals                                                                                                                                       
type = <optimized out>                                                                                                                                  
__PRETTY_FUNCTION__ = "__pthread_mutex_lock"                                                                                                            
id = <optimized out>                                                                                                                                    
(gdb) print session                                                                                                                                     
No symbol "session" in current context.                                                                                                                 
(gdb) 
```

line 622 of my session.c is pthread_mutex_lock as below, have taken some bugfix from the newer versions.
```

    if ((session->side == NC_SERVER) && (session->flags & NC_SESSION_CALLHOME)) {
        /* CH LOCK */
        pthread_mutex_lock(session->opts.server.ch_lock);
    }
```
Any more information needed, please let me know.

Thanks

**Comment by michalvasko** [comment]
Hi,
I think *libnetconf2* commit `5e0edd89` may have fixed this problem.

**Comment by jiantuo** [comment]
Hi Michal,

It is different from #518 , which has been fixed via coommit `5e0edd89`.
This time libnetconf2 hanged in an earlier stage.
You could check the backtrace, it is different too.

Actually, I rasied #518 , and your commit have fixed netopeer2-server crash in that case.

**Comment by michalvasko** [comment]
Okay, then I cannot help you. I will not waste time finding the right commit or debugging a more than 2 year old version of our library.

**Comment by jiantuo** [comment]
Hi Michal,

Could you please give some guide, if we debug this issue ourself?

Thanks

**Comment by michalvasko** [comment]
Well, firstly compile *libnetconf2* in the default `Debug` build target so you can see all the symbols and code lines. Then you can simply try `valgrind` that may print all the information that is required. Or, use `gdb` just like you did and try to learn the cause of the crash yourself.

**Comment by jiantuo** [comment]
Hi Michal,

I tried with the latest legacy branch, still met crash issue.

```root@de33de949cf4:/# sysrepod -v
sysrepod - sysrepo daemon, version 0.7.9

root@de33de949cf4:/# netopeer2-server -V
netopeer2-server 0.7.18
compile time: Sep  1 2020, 10:40:49
root@de33de949cf4:/#
```

Please have a check:
```
Thread 5 "netopeer2-serve" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7faf265f0700 (LWP 76)]
0x00007faf2eef0d28 in nc_session_free (session=0x7faf14006670, data_free=0x559c09d67e34 <free_ds>) at /opt/dev/libnetconf2/src/session.c:778
778                     for (siter = session->ti.libssh.next; siter->status != NC_STATUS_RUNNING; siter = siter->ti.libssh.next) {
(gdb) bt
#0  0x00007faf2eef0d28 in nc_session_free (session=0x7faf14006670, data_free=0x559c09d67e34 <free_ds>) at /opt/dev/libnetconf2/src/session.c:778
#1  0x0000559c09d696fe in np2srv_del_session_clb ()
#2  0x0000559c09d6aef8 in worker_thread ()
#3  0x00007faf2f4206db in start_thread (arg=0x7faf265f0700) at pthread_create.c:463
#4  0x00007faf2e80da3f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) info locals
r = 1
i = 9999
rpc_locked = 1
sock = -1
connected = 1
multisession = 1
siter = 0x0
contiter = 0x4
rpl = 0x0
child = 0x7faf1c0045f0
close_rpc = 0x7faf265effc0
ietfnc = 0x7faf265efe90
p = 0x0
__func__ = "nc_session_free"
(gdb)
```

**Comment by michalvasko** [comment]
Hi,
based on this I see that you have not applied the commit `5e0edd89` because it changes this `for` loop into a `while`. Please stop wasting my time.

**Comment by jiantuo** [comment]
Hi Michal，sorry for the missed commit.

It is the latest legacy docker version, didn't expect it . 
git logs are as below
```
root@de33de949cf4:/opt/dev/libnetconf2# git log -2
commit b21e8674f471d188008094a3c16823553929b8c4 (HEAD -> legacy, origin/legacy)
Author: Jan Kundr<C3><A1>t <jan.kundrat@cesnet.cz>
Date:   Mon Apr 20 15:45:08 2020 +0200

    test CHANGE use ECDSA keys instead of deprecated DSA (#222)
```

I take commit `5e0edd89` now, please check again:
```
[2020/09/09 13:17:49.075321, 1] ssh_socket_exception_callback:  Socket exception callback: 1 (0)
[2020/09/09 13:17:49.075379, 1] ssh_socket_exception_callback:  Socket error: disconnected
[ERR]: Session 12: SSH channel poll error (Socket error: disconnected).
[INF]: Session 12: thread 4 event session terminated.
[ERR]: Session 11: SSH channel poll error (Socket error: disconnected).
[INF]: Session 11: thread 2 event session terminated.
[INF]: Generated new event (netconf-session-end).
[INF]: Generated new event (netconf-session-end).
[INF]: Call Home client "test_ssh_ch_client" session terminated, reconnecting...
[INF]: Trying to connect via IPv4.

Thread 4 "netopeer2-serve" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffb35650700 (LWP 103)]
0x00007ffb396c8d28 in nc_session_free (session=0x7ffb1c0071d0, data_free=0x55aa61bb0e34 <free_ds>) at /opt/dev/libnetconf2/src/session.c:778
warning: Source file is more recent than executable.
778                         siter = session->ti.libssh.next;
(gdb) bt
#0  0x00007ffb396c8d28 in nc_session_free (session=0x7ffb1c0071d0, data_free=0x55aa61bb0e34 <free_ds>) at /opt/dev/libnetconf2/src/session.c:778
#1  0x000055aa61bb26fe in np2srv_del_session_clb ()
#2  0x000055aa61bb3ef8 in worker_thread ()
#3  0x00007ffb39bf86db in start_thread (arg=0x7ffb35650700) at pthread_create.c:463
#4  0x00007ffb38fe5a3f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) info locals
r = 1
i = 9999
rpc_locked = 1
sock = -1
connected = 0
multisession = 1
siter = 0x0
contiter = 0x4
rpl = 0x0
child = 0x7ffb2c004190
close_rpc = 0x7ffb3564ffc0
ietfnc = 0x7ffb3564fe90
p = 0x0
__func__ = "nc_session_free"
(gdb) list
773                     /* there are still multiple sessions, keep the ring list */
774                     siter->ti.libssh.next = session->ti.libssh.next;
775                 }
776                 /* change nc_sshcb_msg() argument, we need a RUNNING session and this one will be freed */
777                 if (session->flags & NC_SESSION_SSH_MSG_CB) {
778                         siter = session->ti.libssh.next;
779                         while (siter && siter->status != NC_STATUS_RUNNING) {
780                         if (siter->ti.libssh.next == session) {
781                             ERRINT;
782                             break;
(gdb) list 760
755
756                         free(siter);
757                     } while (session->ti.libssh.next != session);
758                 }
759                 /* remember sock so we can close it */
760                 sock = ssh_get_fd(session->ti.libssh.session);
761                 if (connected) {
762                     ssh_disconnect(session->ti.libssh.session);
763                     sock = -1;
764                 }
(gdb)
765                 ssh_free(session->ti.libssh.session);
766             } else {
767                 /* remove the session from the list */
768                 for (siter = session->ti.libssh.next; siter->ti.libssh.next != session; siter = siter->ti.libssh.next);
769                 if (session->ti.libssh.next == siter) {
770                     /* there will be only one session */
771                     siter->ti.libssh.next = NULL;
772                 } else {
773                     /* there are still multiple sessions, keep the ring list */
774                     siter->ti.libssh.next = session->ti.libssh.next;
(gdb)
775                 }
776                 /* change nc_sshcb_msg() argument, we need a RUNNING session and this one will be freed */
777                 if (session->flags & NC_SESSION_SSH_MSG_CB) {
778                         siter = session->ti.libssh.next;
779                         while (siter && siter->status != NC_STATUS_RUNNING) {
780                         if (siter->ti.libssh.next == session) {
781                             ERRINT;
782                             break;
783                         }
784                         siter = siter->ti.libssh.next;
(gdb) 
```

**Comment by michalvasko** [comment]
Oh, you are using `docker`. If you notice the warning
```
warning: Source file is more recent than executable.
```
it means that you are actually not running on this version but still on the old one without the fix. I have experienced something similar. The library was used only after I manually compiled and installed it with the bugfix in the image itself so I suggest you do the same. I have not figured out the exact problem but it made me dislike `docker` even more.
